### DISCUSSION TITLE
How to have a single storybook for multiple libraries?

### DISCUSSION THREAD

**Discussion by stefan-schweiger** [nrwl/nx#6215] [General]
Hi everyone, how do you guys handle using storybook with multiple libraries?

In our case we want to create our own ui framework for our applications, which has multiple libraries but we would like them combined in a single storybook.

For now we have 3 libs `shared-ui-inputs` , `shared-ui-containers` and `shared-ui-charts` and for each of those there is a storybook configured. But this means that there are 3 separate storybooks and we never can publish one containing all components/stories.

Do you just live with that or how do you create a single storybook for multiple libraries? Any other ideas how to handle this?

**Comment by twittwer** [comment]
We use a special lib as combined storybook. Nrwl's YouTube channel provides a according tutorial: https://youtu.be/c323HOuFKkA

**Reply by stefan-schweiger** [reply]
@twittwer thanks for the reply :) Are you using compodoc to generate documentation for the Docs addon? If yes, how do you handle either multiple documentation.json files or building everything into one?

**Reply by twittwer** [reply]
Yes, we create one documentation.json as part of the tooling lib.

Maybe this can help as well: https://www.npmjs.com/package/@twittwer/compodoc

**Reply by seanlawrenz** [reply]
@twittwer 

~~That video is out of date as there is no longer a config file to join multiple storybook files.~~

Looks we can achieve the same outcome by using the `main.js` file.

**Reply by twittwer** [reply]
Can you explain what/how you're achieving with `main.js`?

**Reply by testaccount00075433** [reply]
@seanlawrenz is there any repo example with main.js solution?

**Reply by aidanxyz** [reply]
Here is the latest doc on how to achieve it using `main.js`: https://nx.dev/recipes/storybook/one-storybook-for-all

**Comment by asherccohen** [comment]
I am also using a separate lib for shared storybook:
- create a lib for components and design system (lib/UI)
- create a lib with storybook capabilities (libs/shared/storybook)
- import components and write stories in /libs/shared/storybook
- stories can be exported and used in other files of the same lib if you have theme variants 
- inject components as props of a story

It's a bit mechanical but once done it works great.
I can showcase components styled in multiple themes in the same instance and only write stories once.

**Reply by silouanwright** [reply]
I guess my problem with that, is that relationally, I would want the stories to live right next to where the components are defined?

**Comment by GrumpyMeow** [comment]
Hi,
I’ve put much work into getting Storybook and Nx to play nice. I went back and forth between Storybook Composition and a single Shared-Storybook instance. My conclusion is that the only implementation which aligns with Angular and NX is based on Storybook Composition.

This as I had to provide different providers/settings per project for example AngularMaterial form-field default layout (outline, legacy, etc). This as I have to support multiple Material themes (multiple brands).
I also had to configure different style-prefix-urls for the imports from within scss.

Having a single large Storybook is much much faster compared to multiple Storybook instances. In our situation it takes over 2 hours to build-storybook when all projects were affected. This “cost” makes that it nears the breakeven of to (non-)use Storybook.
I’m sure that performance will improve with Storybook v7 in the near future. 

I think my current implementation aligns with the one of Nx/Katerina. One thing i did implement is the automatic generation of the “refs” needed for Composition by reading the Nx-workspace configuration.
